ETT-752 sbcr project

moseshll opened the pull request:
The SBCR UI is a combination of the Core and Crown Copyright projects. No attempt has been made to reduce duplication between the new code and the existing UI components (*.tt bits), since we don't have sufficient test coverage, in particular no way to Playwright the thing.

The new `Project::SBCR` subclass has 100% test coverage, most of that being in the `ValidateSubmission` method which does a number of checks on user-entered review data. Many of these were written (based on checks in other projects) and not needed for SBCR so they are applied toward better coverage in Project.t (which holds the default set of validations for monographs/state gov docs projects) mostly. More could be done, either by copying the tests to other projects, or by extracting out common validations to the `Project` class. That's out of scope.

CRMS::Entitlements (also 100% under test) was added because the `CRMS.pm` methods are gross wrappers around SQL-as-API and I wanted something not horrible for the `SBCR.pm` tests. The comments indicate this may be a temporary class name because `CRMS::Rights` is currently taken... but I have to confess `Entitlements` is growing on me.

The `UNIQUE` constraint on `crms.rights.{attr/reason}` has already been applied in production and htdev.

moseshll committed:
ETT-752 CRMS project interface - subject based copyright review (SBCR)
  - Add `Project` subclass and views for UI testing on dev-2
  - Still TODO:
    - Unit tests
    - Input validation

moseshll committed:
prediction loader (busy animation) should be inline-block and not just block when displayed

moseshll committed:
- Add a number of data validations to SBCR project
- Add parameter extraction of all fields in SBCR UI
- Review data format
- Extract out `partial/rights.tt` which should be reusable across all projects

moseshll committed:
Add a couple more SBCR data validations

moseshll committed:
User data import fixes

moseshll committed:
Add `CRMS::Entitlements` intended to eventually replace the attributes/reasons/rights routines in CRMS.pm
with a more structured class that is not just a wrapper around SQL queries:
  - `TranslateAttr`
  - `TranslateReason`
  - `TranslateRights`
  - `AllCRMSRights`
  - `Rights`
  - `AllAssignableRights`
  - `GetAttrReasonFromCode`
  - `TranslateAttrReasonFromCode`
  - `GetCodeFromAttrReason`
  - `GetAttrReasonCode`

This class is fully tested. For now it will only be used in the SBCR module and its tests.

moseshll committed:
- SBCR module uses `CRMS::Entitlements`
- SBCR module tests, first pass

moseshll committed:
- Add `UNIQUE` index to rights table and note about it in `Entitlements.pm`
- Eager load only when populating the singleton

moseshll committed:
ExtractReviewData uses extract_parameters to take advantage of whitespace trimming

moseshll committed:
FormatReviewData test

moseshll committed:
100% coverage on SBCR.pm and Entitlements.pm modules

moseshll committed:
Move unused review data validations from SBCR to projects that use them

aelkiss commented on cgi/partial/sbcr_form.tt:
```diff
@@ -0,0 +1,341 @@
+      </td>
+    </tr>
+  </table>
+  [% IF expert && importUser %]
```
If this is section is copied from somewhere else, I say leave it; if it's new, it might be worth extracting as a partial

moseshll replied:
Yes, this is now in its own `import_user.tt` partial with note on possible future simplification.

aelkiss commented on cgi/partial/rights.tt:
Lots of logic in this partial; possible to extract some of it?

moseshll replied:
We are using our own rights descriptions originally written by AKZ, these have been unmaintained. Experiment with just using what's in ht_rights. Kristina gave this the thumbs-up.

Also, added a utility for pairing up rights which makes the rights.tt logic prettier.

May split out the first half of the partial into its own but I feel less need to do so now that the logic has been simplified and we have a good idea what this will look like when more projects adopt the new UI partitioning.

aelkiss commented on cgi/partial/sbcr_form.tt:
```diff
@@ -0,0 +1,341 @@
+</div>
+</form>
+
+<script>
```
Would the static JS make sense to extract elsewhere, and just have the review data as JSON here in the template?

moseshll replied:
I don't think so. Each project has its own version of many of these routines but they will tend to differ in subtle ways. For example the list of data entry fields that may need to be synchronized (checkboxes that can slightly rearrange the UI, AJAX calls to guide rights selection, etc). To do it properly would require revisiting the commonalities and differences across all projects, and that is out of scope. Really needs the scaffolding of something beyond just unit tests.

aelkiss commented on cgi/partial/sbcr_form.tt:
```diff
@@ -0,0 +1,341 @@
+addEvent(window, 'load', partial_sbcr_form_mainWindowLoad);
+
+function partial_sbcr_form_mainWindowLoad(e) {
+  [% IF importUser %]
```
Could we have `importUser` as a JSON block as well, and then just do something like this (the syntax is probably wrong but you get the idea); then all the JS could be in a static file

```javascript
var importUser = [% importUserJson %];

if(importUser != null) {
  insertUserReviewData(importUser);
}
```

moseshll replied:
I'd love to get rid of this local JS, but as mentioned all projects have some variation on these routines with their own tailored differences. Keeping them here keeps them sequestered to the one spot they're used.

aelkiss reviewed (commented):
Generally the perl side is looking pretty good but I wonder if we might be able to clean up a bit in the templates.

* individual tests are easy to follow, although I didn't read through each one in detail
* `SBCR.pm` and `Entitlements.pm` look fine to me.
* I don't know that we have another name for the combination of attr/reason that 'entitlement' represents other than "attribute/reason combination" -- as "rights" typically to me implies the combination of the item ID and the attribute & reason codes. 'entitlement' seems as reasonable as anything.
* I had some comments on the templates; if they're copied from other existing things that might make it more involved to address the comments. If they're new for this project, maybe it's worth addressing now.

moseshll committed:
Remove redundant call to crms.Rights

moseshll committed:
Add `use CRMS::Entitlements` to SBCR.pm

moseshll committed:
- Fix copypasta in JS popRenewalDate function that fortunately should not have led to aberrant behavior
- Add comment about the rights way to do the rights <-> id mapping in JS code

moseshll committed:
Add notes on how to reuse rights.tt

moseshll committed:
Remove prediction loader img left over from Commonwealth UI and effectively commented out

moseshll committed:
- Extract out the notes category menu and textarea into its own <div> partial used by SBCR
- Should be usable by other projects by swapping in an `INCLUDE` in each project's main form

moseshll committed:
Split user review data import logic off into import_user.tt partial

moseshll committed:
Experiment: use attr/reason descriptions from ht_rights rather than our own poorly maintained version

moseshll committed:
- Add CRMS.pm `array_to_pairs` utility to simplify logic in rights.tt partial

moseshll committed:
Make CRMS::array_to_pairs comments clearer

aelkiss reviewed (approved):
Additional changes & reasons for not addressing other items now all make sense to me.

moseshll committed:
Save and restore SBCR project ADD value when pub date checkbox is toggled so user-entered data is not lost.

moseshll committed:
Version bump

moseshll committed:
Replace expertDetails.tt with new expertDetails_sbcr.tt viewable by non-experts, in SBCR project.

moseshll committed:
Merge branch 'main' into ETT-752_sbcr_project
